qiskit/algorithms/minimum_eigen_solvers/vqe.py

```diff
@@ -208,8 +238,7 @@ def expectation(self) -> ExpectationBase:
 
     @expectation.setter
     def expectation(self, exp: ExpectationBase) -> None:
-        self._set_expectation(exp)
-        self._user_valid_expectation = self._expectation is not None
+        self._expectation = exp
```

Comment:
I tried to look over the code but I am not sure the new code behaves as was intended and done how it was before. When the factory was used to select/create an expectation it was with the understanding that it would do so based on the operator/backend. Whereas we always use the expectation object a user explicitly provides - which may fail if there is some mismatch - the default/factory way was supposed to select one accordingly even as things changed. Hence there were a couple of fields to track this in the past since simply setting the expectation into a single variable we had no idea if thing started out None (or we set back to None) so this would be done via the factory which could discard any expectation that was there for anew match, versus having to keep/use it since its the one a user gave us explicitly

Reply:
So if I read the code correctly, before we did: if the user sets any `expectation`, use that. If not, set it via the expectation factory. I think it should still be the same now, but maybe I didn't get the previous behavior right?

I'm didn't understand why there were additional arguments to keep track of that, in the end all you have to know is whether the `expectation` is provided or is `None`, no?

Reply:
In the past we stored the expectation in some internal instance. Though that maybe None to start with by the time we set it with some expectation, so it becomes non-None, we needed the extra info to remember the origin of the non-None value. If it was from the user we should just use it; if it came from the factory, since we say its auto selected base on operator etc, then if the user re-uses the VQE instance, but changes something that would affect which expectation we should use we should get one from the factory that supports the new configuration. So when you create a VQE from scratch indeed looking at the _expectation for None would suffice - but once it used that will be set; and once it is, and we go to re-use VQE, looking at it for None now longer worked.

Reply:
I see! But that can be easily fixed by doing 
```python
# in compute_minimum
if self._expectation is None:
     expectation = # set via factory
else:
     expectation = self._expectation
     
# use expectation instead of self._expectation

```

Reply:
Right, if it can be set like that and used/passed around as needed rather than relying on self._expectation, which remains whatever the user set, then that indeed works.

Reply:
Actually, that didn't work. I reverted to the old behavior (but a little different variables).